Assets/Script/Gameplay/TextureManager.cs

```diff
@@ -0,0 +1,125 @@
+    /// A point of integration between GamePlay state
+    /// and venue materials
+    /// </summary>
+    public class TextureManager : GameplayBehaviour
+    {
+
+        // Smoothing factor (adjust to taste)
+        [Range(0.0f, 0.95f)]
+        public float fftSmoothingFactor = 0.8f;
+        [Range(0.0f, 0.95f)]
+        public float waveSmoothingFactor = 0.5f;
+
+        private Texture2D _sourceIcon;
+        private Texture2D _albumCover = null;
+        private Texture2D _soundTexture = null;
+        private float[] _fft = new float[FFT_SIZE * 2];
+        private float[] _wave = new float[FFT_SIZE * 2];
+        private float[] _prevFft = new float[FFT_SIZE * 2];
+        private float[] _prevWave = new float[FFT_SIZE * 2];
+
+        private static int _soundTexId = Shader.PropertyToID("_Yarg_SoundTex");
+        private static int _sourceIconId = Shader.PropertyToID("_Yarg_SourceIcon");
+        private static int _albumCoverId = Shader.PropertyToID("_Yarg_AlbumCover");
+
+        private const double MIN_DB = -100.0;
+        private const double MAX_DB = -30.0;
+        private const double DB_RANGE = MAX_DB - MIN_DB;
+        private const int FFT_SIZE_LOG = 9 /* aka log2(512) */;
+        private const int FFT_SIZE = 1 << FFT_SIZE_LOG;
+
+        private void Start()
+        {
+            _sourceIcon = SongSources.SourceToIcon(GameManager.Song.Source).texture;
+        }
+
+        protected Texture2D getAlbumArt()
+        {
+            if (_albumCover == null)
+            {
+                var image = GameManager.Song.LoadAlbumData();
+                _albumCover = image.LoadTexture(false);
+            }
+            return _albumCover;
+        }
+
+        protected Texture2D getSoundTexture()
+        {
+            if (_soundTexture == null)
+            {
+                // first row is FFT data
+                // second is waveform data
+                _soundTexture = new Texture2D(FFT_SIZE, 2, TextureFormat.R8, false);
+            }
+            return _soundTexture;
+        }
+
+        public void processMaterial(Material m)
+        {
+            if (m.HasTexture(_sourceIconId))
+            {
+                m.SetTexture(_sourceIconId, _sourceIcon);
+            }
+            if (m.HasTexture(_soundTexId))
+            {
+                m.SetTexture(_soundTexId, getSoundTexture());
+            }
+            if (m.HasTexture(_albumCoverId))
+            {
+                m.SetTexture(_albumCoverId, getAlbumArt());
+            }
+        }
+
+        private async void UpdateFFT()
+        {
+            if (_soundTexture == null)
+            { 
+                return; 
+            }
+            var pd = _soundTexture.GetPixelData<Byte>(0);
+
+            await UniTask.RunOnThreadPool(() =>
+            {
+                GameManager.GetMixerFFTData(_fft, FFT_SIZE_LOG, false);
+                GameManager.GetMixerSampleData(_wave);
+
+                for (int i = 0; i < FFT_SIZE; ++i)
```

Comment:
I believe the ShaderToy JS actually only uses the first 512 bins of the 1024 returned for an FFT2048

```suggestion
                for (int i = 0; i < FFT_SIZE / 2; ++i)
```